state/globalclock/updater.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package globalclock
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/core/globalclock"
+	"github.com/juju/loggo"
+	mgo "gopkg.in/mgo.v2"
+)
+
+var (
+	logger = loggo.GetLogger("juju.state.globalclock")
+
+	globalEpoch = time.Unix(0, 0)
+)
+
+// NewUpdater returns a new Updater using the supplied config, or an error.
+//
+// Updaters do not need to be cleaned up themselves, but they will not function
+// past the lifetime of their configured *mgo.Session.
+func NewUpdater(config UpdaterConfig) (*Updater, error) {
+	if err := config.validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	updater := &Updater{config: config}
+	t, err := updater.ensureClock()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	updater.time = t
+	return updater, nil
+}
+
+// Updater provides a means of updating the global clock time.
+//
+// Updater is not goroutine-safe.
+type Updater struct {
+	config UpdaterConfig
+	time   time.Time
+}
+
+// Advance adds the given duration to the global clock, ensuring
+// that the clock has not been updated concurrently.
+//
+// Advance will return ErrConcurrentUpdate if another updater
+// updates the clock concurrently. In this case, the updater
+// will refresh its view of the clock, and the caller can
+// attempt Advance later.
+//
+// If Advance returns any error other than ErrConcurrentUpdate,
+// the Updater should be considered invalid, and the caller
+// should obtain a new Updater. Failing to do so could lead
+// to non-monotonic time, since there is no way of knowing in
+// general whether or not the database was updated.
+func (u *Updater) Advance(d time.Duration) error {
+	if d < 0 {
+		return errors.NotValidf("duration %s", d)
+	}
+	coll := u.config.Session.DB(u.config.Database).C(u.config.Collection)
+	new := u.time.Add(d)
+	if err := coll.Update(matchTimeDoc(u.time), setTimeDoc(new)); err != nil {
+		if err == mgo.ErrNotFound {
```

Comment:
The delay is purely there to avoid unnecessary database reads and writes. Maybe it's overkill. The worker/lease code would work just as well without it, and we'd have less of a delay in taking over leadership. Note though that it only takes effect when the leader fails, so a short delay shouldn't be unexpected.

We don't *need* sub-second precision, but it keeps the code simple and moves the global time along as far as we safely/correctly can. Is there a concern with this?

I don't understand what you're saying about convergence.